hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/AltoChannelConnector.java

```diff
@@ -0,0 +1,215 @@
+ * with the connection.
+ */
+public final class AltoChannelConnector {
+    private final UUID clientUuid;
+    private final TcpClientConnection connection;
+    private final List<Integer> altoPorts;
+    private final ExecutorService executor;
+    private final ChannelInitializer channelInitializer;
+    private final ChannelCreator channelCreator;
+    private final ILogger logger;
+    private final Channel[] altoChannels;
+    private final AtomicInteger remaining;
+    private volatile boolean failed;
+
+    public AltoChannelConnector(UUID clientUuid,
+                                TcpClientConnection connection,
+                                List<Integer> altoPorts,
+                                ExecutorService executor,
+                                ChannelInitializer channelInitializer,
+                                ChannelCreator channelCreator,
+                                LoggingService loggingService) {
+        this.clientUuid = clientUuid;
+        this.connection = connection;
+        this.altoPorts = altoPorts;
+        this.executor = executor;
+        this.channelInitializer = channelInitializer;
+        this.channelCreator = channelCreator;
+        this.logger = loggingService.getLogger(AltoChannelConnector.class);
+        this.altoChannels = new Channel[altoPorts.size()];
+        this.remaining = new AtomicInteger(altoPorts.size());
+    }
+
+    /**
+     * Initiates the connection attempts.
+     * <p>
+     * This call does not block.
+     */
+    public void initiate() {
+        logger.info("Initiating connection attempts to Alto channels running on ports "
+                + altoPorts + " for " + connection.toLessDescriptiveString());
+        String host = connection.getRemoteAddress().getHost();
+        int i = 0;
+        for (int port : altoPorts) {
+            int index = i++;
+            executor.submit(() -> connect(host, port, index));
+        }
+    }
+
+    private void connect(String host, int port, int index) {
+        if (connectionFailed()) {
+            // No need to try to connect if one of the channels
+            // or the connection itself is closed/failed.
+            logger.warning("The connection to Alto channel on port "
+                    + port + " for " + connection.toLessDescriptiveString()
+                    + " will not be made as either the connection or one of "
+                    + "the Alto channel connections has failed.");
+            return;
+        }
+
+        logger.info("Trying to connect to Alto channel on port " + port + " for "
+                + connection.toLessDescriptiveString());
+
+        Channel channel = null;
+        try {
+            Address address = new Address(host, port);
+            channel = channelCreator.create(address, connection, channelInitializer);
+            writeAuthenticationBytes(channel);
+            onSuccessfulChannelConnection(channel, index);
```

Comment:
At this point we are not sure if the uuid bytes are written into the network and received by the server successfully but this method assumes that it is success.

Reply:
Yes, that change will be in the next PR. Like I said above, it is not done in this PR, as it was part of another task